lib/src/request.dart

```diff
@@ -175,11 +175,17 @@ class Request extends Message {
           requestedUri, 'requestedUri', 'may not have a fragment.');
     }
 
-    if (this.handlerPath + this.url.path != this.requestedUri.path) {
+    // Notice that because relative paths must encode colon (':') as %3A we
+    // cannot actually combine this.handlerPath and this.url.path, but we can
+    // compare the pathSegments. In practice exposing this.url.path as a Uri
+    // and not a String is probably the underlying flaw here.
+    final pathSegments = Uri(path: this.handlerPath).pathSegments.join('/') +
```

Comment:
Should this be `Uri(path: handlerPath).pathSegments.followedBy(url.pathSegments).join('/')`? Do we intend for the last segment of `handlerPath` to get merged with the first segment in `url` without a slash between them?

Reply:
@jonasfj - is this still something you want to look at?

Reply:
Sorry, I'm slow... thanks for the ping :)

```dart
    final pathSegments = Uri(path: this.handlerPath)
        .pathSegments
        .followedBy(this.url.pathSegments)
        .join('/');
```
Won't work because `handlerPath` always ends in `/`.

Hence, if `handlerPath` is `/static/` and `url` is `dir/file.html` we would get `static//dir/file.html` instead of `/static/dir/file.html` (which is the request path). I suppose `static/` gives rise to an empty path segment to preserve trailing slash.

-----

If I understand it right `handlerPath` is always a prefix of `requestedUri.path`. What we are trying to do is check that `this.url.path` is the remainder of the path. This impossible because making a suffix of `requestedUri.path` into a relative path encoded with `Uri` causes encoding of colon (which is destructive, we can't tell the difference between colon and `%3A`).

But since this is just an assertion, not something that should happen at all. I think we can ignore the case where `requestedUri` contains an encoded character and `uri` contains the decoded counter part. It's technically possible to create this invariant violation with the `Request` constructor.

So I essentially opted to relax the invariant check by always decoding before we compare. I think this is better than having valid requests cause internal errors.

Reply:
We could also drop this invariant check completely :see_no_evil: 

But I think it can catch some serious mistakes, and that it can still do this. Even if we relaxed it a bit.

Reply:
I think we want to be using `package:path` or maybe `IterableEquality`. We already depend on both `path` and `collection` packages. I'm very nervous about this manual string concatenation.

We can get the path segments without the last empty one, and then append `this.url.pathSegments`.

```dart
    final pathSegments = Uri(path: handlerPath)
        .pathSegments
        .takeWhile((s) => s.isNotEmpty)
        .followedBy(this.url.pathSegments);
```

Then compare with either

```dart
    if (p.url.joinAll(pathSegments) != p.url.joinAll(this.requestedUri.pathSegments)) {
```

or

```dart
    if (const IterableEquality().equals(pathSegments, this.requestedUri.pathSegments)) {
```

Reply:
>         .takeWhile((s) => s.isNotEmpty)
Are you sure a path with multiple slashes `/folder///subfolder/file.txt` won't have empty segments.

Reply:
I do think it would be unusual to have a `handlerPath` with empty segments...